File: iosApp/iosApp/Modules/CurrencyReceive/CurrencyReceiveInteractor.swift
// Created by web3d4v on 13/06/2022.
// Copyright (c) 2022 Sons Of Crypto.
// SPDX-License-Identifier: MIT

import Foundation
import web3walletcore

protocol CurrencyReceiveInteractor: AnyObject {
    func receivingAddress(network: Network, currency: Currency) -> String
}

final class DefaultCurrencyReceiveInteractor {
    private let networksService: NetworksService
    
    init(networksService: NetworksService) {
        self.networksService = networksService
    }
}

extension DefaultCurrencyReceiveInteractor: CurrencyReceiveInteractor {

    func receivingAddress(network: Network, currency: Currency) -> String {
        // TODO: Review this in the future when currencies may have its own receiving address
        // for now this is not an issue though.
        guard let address = try? networksService.wallet(network: network)?.address() else {
            fatalError("Unable to read wallet address from network \(network.name)")
        }
        return address.toHexStringAddress().hexString
    }
}

File: web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/modules/currencyReceive/CurrencyReceiveInteractor.kt
package com.sonsofcrypto.web3walletcore.modules.currencyReceive

import com.sonsofcrypto.web3lib.services.networks.NetworksService
import com.sonsofcrypto.web3lib.types.Currency
import com.sonsofcrypto.web3lib.types.Network
import com.sonsofcrypto.web3lib.types.toHexStringAddress

interface CurrencyReceiveInteractor {
    fun receivingAddress(network: Network, currency: Currency): String?
}

class DefaultCurrencyReceiveInteractor(
    val networksService: NetworksService
): CurrencyReceiveInteractor {

    override fun receivingAddress(network: Network, currency: Currency) =
        networksService.wallet(network)?.address()?.toHexStringAddress()?.hexString
}